runtime/src/accounts_db.rs

```diff
@@ -1525,27 +1529,69 @@ impl AccountsDB {
         hasher.result()
     }
 
-    fn accumulate_account_hashes(mut hashes: Vec<(Pubkey, Hash)>) -> Hash {
-        let mut sort = Measure::start("sort");
+    fn accumulate_account_hashes(hashes: Vec<(Pubkey, Hash, u64)>) -> Hash {
+        let (hash, ..) = Self::do_accumulate_account_hashes_and_capitalization(hashes, false);
+        hash
+    }
+
+    fn accumulate_account_hashes_and_capitalization(
+        hashes: Vec<(Pubkey, Hash, u64)>,
+    ) -> (Hash, u64) {
+        let (hash, cap) = Self::do_accumulate_account_hashes_and_capitalization(hashes, true);
+        (hash, cap.unwrap())
+    }
+
+    fn do_accumulate_account_hashes_and_capitalization(
+        mut hashes: Vec<(Pubkey, Hash, u64)>,
+        calculate_cap: bool,
+    ) -> (Hash, Option<u64>) {
+        let mut sort_time = Measure::start("sort");
         hashes.par_sort_by(|a, b| a.0.cmp(&b.0));
-        sort.stop();
-        let mut hash_time = Measure::start("hash");
+        sort_time.stop();
 
-        let fanout = 16;
+        let mut sum_time = Measure::start("cap");
+        let cap = if calculate_cap {
+            Some(hashes.iter().fold(0, |acuum: u64, (_, _, lamports)| {
+                acuum.checked_add(*lamports).unwrap()
+            }))
+        } else {
+            None
+        };
+        sum_time.stop();
 
+        let mut hash_time = Measure::start("hash");
+        let fanout = 16;
         let res = Self::compute_merkle_root(hashes, fanout);
-
         hash_time.stop();
-        debug!("{} {}", sort, hash_time);
 
-        res
+        debug!("{} {} {}", sort_time, hash_time, sum_time);
+
+        (res, cap)
+    }
+
+    pub fn account_balance_for_capitalization(
+        lamports: u64,
+        owner: &Pubkey,
+        executable: bool,
+    ) -> u64 {
+        let is_specially_retained = (solana_sdk::native_loader::check_id(owner) && executable)
```

Comment:
This `executable` is needed to be unblocked from #11928

Reply:
@sakridge  sorry for rather intrusive revert of your changes (65342994f05b309b92b137533dbc48f83f905708). But, I think this account.owner-based condition is just enough for capitalization calculation purpose, instead of carrying around `pubkeys`. That's because I want to avoid adding bank fields unless really needed. And new check is state-less, so less risk of introducing bugs. And lastly, this will be faster or equivalent in terms of performance, compared to 1 lookup of `HashSet` for each account.

Was there any concern I'm missing for choosing using `builtin_programs` and `genesis_cap_exempt_ids`?.

As for deprecated storage accounts/rewards pool, I'm going to address them first. It's living only in testnet at this moment.

Reply:
No, that's great. Definitely better to avoid adding bank fields if we can get away with it. I wasn't actually sure about the exact conditions where capitalization wasn't applied.

Reply:
> And lastly, this will be faster or equivalent in terms of performance, compared to 1 lookup of HashSet for each account.

Might be debatable, because this sysvar::check_id has to do a compare with all existing sysvars. The hashset just has to hash the account value and then check the table to see if it exists in the set. It would probably touch a lot less memory.

Reply:
I see. That `calculate_capitilzation` function didn't exist when I started this PR and I missed that it was added.

Reply:
> Might be debatable, because this sysvar::check_id has to do a compare with all existing sysvars.

Fortunately, this is not true. We're comparing `account.owner` not indivisual actual sysvar addresses. Maybe you're referring to this: https://github.com/solana-labs/solana/blob/8bcc8edf47b556c7177bb9058dacc62dd101c575/sdk/src/sysvar/mod.rs#L21-L31. But we're intead doing this: https://github.com/solana-labs/solana/blob/8bcc8edf47b556c7177bb9058dacc62dd101c575/sdk/macro/src/lib.rs#L48-L50

So, simple byte-wite comparison of `Pubkey` (note: this is `derive(Copy)`).

And, because almost all  of comparison is negative, this is super friendly for branch prediction, bailing out after few prefix bytes. And the required data is a single `Pubkey` or two if you include `NativeLoader`, so this will be cpu-cache super friendly as well. :)

At least, HashSet requires hashing the whole Pubey, without any short-circuitting. So, I think this will be faster in theory. I haven't measured it, though... ;)

Reply:
Yea, I mean you have to do a load for each of those sysvar Pubkey which even if you only read the 1st byte, you still do a load on a fresh cache line, and each pubkey will likely be on different cache lines. So you have to 10 loads even if its' just looking at the first byte of the pubkey.

To do the hash of the whole pubkey means you only need to load in a single 64-byte cache line since the pubkey will be contiguous in memory. The arithmetic instructions to do the hash are then essentially free. Then you need 1 more load at the HashSet table offset.

So 2 loads vs. 10.

Reply:
> > HashSet requires hashing the whole Pubey, without any short-circuitting.

> The arithmetic instructions to do the hash are then essentially free.

Oh, SipHash is that pipeline-freindly and fast? I never knew..

And, thanks for explaining.

So just to be sure, the 10 loads is for `is_sysvar_id`, which we won't use here.

We're going to use `check_id`, which should cost 2 loads according to your explanation (account's `owner` pubkey itself and the literal `Sysvar1111111111111111111111111111111111111`).

Reply:
great. thanks